ext/http/response_body.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+use std::borrow::Cow;
+use std::cell::RefCell;
+use std::future::Future;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::task::Waker;
+
+use deno_core::error::bad_resource;
+use deno_core::error::AnyError;
+use deno_core::futures::FutureExt;
+use deno_core::AsyncRefCell;
+use deno_core::AsyncResult;
+use deno_core::BufView;
+use deno_core::CancelHandle;
+use deno_core::CancelTryFuture;
+use deno_core::RcRef;
+use deno_core::Resource;
+use deno_core::WriteOutcome;
+use hyper1::body::Body;
+use hyper1::body::Frame;
+use hyper1::body::SizeHint;
+
+#[derive(Clone, Debug, Default)]
+pub struct CompletionHandle {
+  inner: Rc<RefCell<CompletionHandleInner>>,
+}
+
+#[derive(Debug, Default)]
+struct CompletionHandleInner {
+  complete: bool,
+  success: bool,
+  waker: Option<Waker>,
+}
+
+impl CompletionHandle {
+  pub fn complete(&self, success: bool) {
+    let mut mut_self = self.inner.borrow_mut();
+    mut_self.complete = true;
+    mut_self.success = success;
+    if let Some(waker) = mut_self.waker.take() {
+      drop(mut_self);
+      waker.wake();
+    }
+  }
+}
+
+impl Future for CompletionHandle {
+  type Output = bool;
+
+  fn poll(
+    self: Pin<&mut Self>,
+    cx: &mut std::task::Context<'_>,
+  ) -> std::task::Poll<Self::Output> {
+    let mut mut_self = self.inner.borrow_mut();
+    if mut_self.complete {
+      return std::task::Poll::Ready(mut_self.success);
+    }
+
+    mut_self.waker = Some(cx.waker().clone());
+    std::task::Poll::Pending
+  }
+}
+
+#[derive(Default)]
+pub enum ResponseBytesInner {
+  /// An empty stream.
+  #[default]
+  Empty,
+  /// A completed stream.
+  Done,
+  /// A static buffer of bytes, sent it one fell swoop.
+  Bytes(BufView),
+  /// A resource stream, piped in fast mode.
+  Resource(bool, Rc<dyn Resource>, AsyncResult<BufView>),
```

Comment:
Why is the first chunk passed separately?

Reply:
Could this block in a weird place if the resource takes a long time to resolve the first buffer?

Reply:
Unlikely. I mean, if you are blocking on the calls required to generate the futures that you are returning, you are already doing something wrong and it'll fail in weird ways in other places. Resources should be able to create their associated read futures quickly and then defer to tokio to let them get polled to completion. If a resource is actually blocking in the synchronous part of the async call, then that resource is definitely busted and we need to fix it ASAP.